File: tests/leetcode/106_test.cpp
// Unit tests for LeetCode Solution 106
#include "../../src/leetcode/106.cpp"

#include <gtest/gtest-param-test.h>
#include <gtest/gtest.h>

#include <string>
#include <tuple>
#include <vector>

#include "../utils/tree_util.hpp"

namespace {
class LeetCode106 : public testing::TestWithParam<std::tuple<std::vector<int>, std::vector<int>, std::vector<int>>> {
 protected:
  leetcode_106::Solution solution;
  LeetCode106() {
  }
  ~LeetCode106() override {
  }
  void SetUp() override {
    std::vector<int> tmp;
    std::tie(inorder, postorder, tmp) = GetParam();
    expected = createBTree(tmp);
  }
  void TearDown() override {
    freeTreeNodes(expected);

    expected = nullptr;
  }
  std::vector<int> inorder;
  std::vector<int> postorder;
  TreeNode* expected;
};

// test_suite_name, test_name
TEST_P(LeetCode106, Test) {
  auto actual = solution.buildTree(inorder, postorder);
  ASSERT_EQ(true, treeEquals(actual, expected));
  freeTreeNodes(actual);
}

// prefix, test_suite_name
INSTANTIATE_TEST_SUITE_P(
    leetcode, LeetCode106,
    testing::Values(std::make_tuple(std::vector<int>{9, 3, 15, 20, 7}, std::vector<int>{9, 15, 7, 20, 3},
                                    std::vector<int>{3, 9, 20, NULL_NODE_VAL, NULL_NODE_VAL, 15, 7})));
}  // namespace

File: tests/leetcode/106_test.py
import pytest

from src.leetcode.s106 import Solution
from src.model import ListNode, TreeNode

dataset = lambda: [
    ([9, 3, 15, 20, 7], [9, 15, 7, 20, 3], [3, 9, 20, None, None, 15, 7]),
    ([-1], [-1], [-1]),
]


@pytest.mark.parametrize("inorder,postorder,expected", dataset())
def test(inorder: list[int], postorder: list[int], expected: list[int]):
    root = Solution().buildTree(inorder, postorder)
    assert TreeNode.toList(root) == expected
